staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go

```diff
@@ -43,44 +53,135 @@ import (
 // therefore transformers using this implementation *must* ensure they allow for frequent key
 // rotation. Future work should include investigation of AES-GCM-SIV as an alternative to
 // random nonces.
-type gcm struct {
-	block cipher.Block
-}
+func NewGCMTransformer(block cipher.Block) (value.Transformer, error) {
+	aead, err := cipher.NewGCM(block)
```

Comment:
Is `aead` safe to reuse and threadsafe? previously we constructed a local one within every call to TransformFromStorage / TransformToStorage, now we reuse the same one for the whole instance. It's not clear to me that is safe.

Reply:
Yup, a member of the Go team confirmed that it was go routine safe / stateless (the std lib code is hard to read because it is mostly assembly).

Reply:
In `master` we already rely on the assumption that `cipher.Block` is go routine safe, so this does not really change anything:

https://github.com/kubernetes/kubernetes/blob/fcf5d23e6818ccfbc92f332af82f31661f8d4f88/staging/src/k8s.io/kms/pkg/encrypt/aes/aes.go#L50-L60

From the discussion in https://github.com/golang/go/issues/41689 (which was a follow up to https://github.com/golang/go/issues/25882), `cipher.NewGCM` is also go routine safe (and it is stateless).

Even without the race detector, `TestGCMUnsafeNonceGen` would detect data corruption (so at least we know it is okay on `linux/amd64` and `darwin/arm64`).

Reply:
I did trace that we were relying on the aes.NewCipher implementation of cipher.Block to be threadsafe, so I agree this PR doesn't change that.

From what I can tell, it does introduce reliance on cipher.NewGCM being threadsafe.

I agree the comments on those go issues and from the go maintainer indicate it is threadsafe, but I would feel *much* better if that was tested/documented on cipher.NewGCM and aes.NewCipher. Perhaps a post-code freeze contribution to go1.21 to solidify what we're relying on here?